### ISSUE TITLE
Change BMIConfig obj

### ISSUE THREAD

**Issue by chemistry-sourabh** [CCI-MOC/ims#28]
Currently the BMIConfig obj reads each field from the config and stores them as a variable.
The Task here is to change these variables to dictionaries that store the entire section.
Also Checking for mandatory fields needs to be added.
This will allow for BMI to use Drivers.

**Comment by sirushtim** [comment]
I'd much rather use oslo.config[1] which enables us to deal with configuration files in a declarative fashion thus saving a lot of time. It automatically has support for checking of mandatory fields too AFAIR.

[1] https://github.com/openstack/oslo.config

**Comment by chemistry-sourabh** [comment]
This would require everybody to learn oslo which I feel is a bad idea. Also oslo is an openstack project does it make sense to use it ? I dont see any real advantage right now.
The only logical change to config would be to convert into json, but this style is much easier to write.

**Comment by sirushtim** [comment]
So, think about it. In the long term, keeping declarative config code would be easier to maintain than having imperative code that would need continuous testing.

Oslo.config although under the openstack umbrella is a python library that's project agnostic and designed to only deal with config files.

Adding a config option with Oslo.config and making it a required/mandatory option is as simple as adding a single line of code which is easy for beginners tbh. Compare that to your method of adding several lines of code and testing for a new config option and furthermore maintaining separate config handling code is only going to waste time in the future IMO.

**Comment by ravisantoshgudimetla** [comment]
This is an interesting discussion and @sirushtim has a valid point but I think using oslo may be an overkill in this scenario. As far as I understood, it has broad set of usages and one of them is reading INI file, others include RBAC policy enforcement but I liked the declarative approach. Why can't we combine both and use yaml for reading configurations(which is much more declarative) and use pyyaml or someother library for reading. Again, this needs an internal discussion and I advise taking votes and proceed further.

**Comment by chemistry-sourabh** [comment]
I kind of like the later approach, but I am kind of confused on the whole declarative thing. Can anyone explain or put some nice link ?

**Comment by sirushtim** [comment]
@ravisantoshgudimetla So, oslo comprises of a set of libraries, each one handling a specific use case. For example, the oslo.config library deals with INI config files. The oslo.policy library deals with RBAC policy enforcement. The suggestion is to not use all the oslo libraries. It is only to use oslo.config here. FWIW, YAML sounds like a nice idea too but that would mean testing and maintaining separate configuration handling code.

@chemistry-sourabh This might help - https://en.wikipedia.org/wiki/Declarative_programming

Also, here is a simple usage of oslo.config as found in the official documentation: http://docs.openstack.org/developer/oslo.config/cfg.html#module-oslo_config.cfg

**Comment by ravisantoshgudimetla** [comment]
@sirushtim Thanks for clarifying. I liked the oslo.config(if it is used alone), in fact it has cli parsing and adding capabilities but again this may be an overkill as we are already thinking of using 'click' for command-line parsing. So, to keep it simple I think of using a yaml file and parse it, let the options be picked up by cli library that we are planning to write. Again, lets see what others think and we will proceed.

**Comment by sirushtim** [comment]
@ravisantoshgudimetla Yeah, the idea is indeed to use oslo.config alone which really can't do much more than dealing with ini files, a very small library in itself. It doesn't seem to have CLI parsing. Where did you see it?

FWIW, OpenStack uses a different library called cliff[1] for command line parsing.

[1] http://docs.openstack.org/developer/cliff/

**Comment by ravisantoshgudimetla** [comment]
It was there in the same link that you pointed out. 

http://docs.openstack.org/developer/oslo.config/cfg.html#loading-config-files.

**Comment by sirushtim** [comment]
@ravisantoshgudimetla Thanks, I just noticed it now. It's an added feature to get the config options from the command line than a config file. I don't think it hurts to have it though. 

I'm mostly concerned about the instrumentation it takes to add a new config variable should we decide to write our own config parsing library. Here's an example from Cinder of what it looks like to add a new config option when we incorporate oslo.config: https://github.com/openstack/cinder/blob/master/cinder/common/config.py#L48-L50.

If we can come up with something similar/easier with yaml files, that'd be cool.

**Comment by chemistry-sourabh** [comment]
@CCI-MOC/bmi-breakers better decide on this as I dont see any progress except the issues piling up.

**Comment by sirushtim** [comment]
@chemistry-sourabh I don't see why this patch should block the progress of the other patches, can you rebase the other patches without this patch? Also, can you make independent pull requests so we'll never be blocked on a certain patch to review another patch.

**Comment by ravisantoshgudimetla** [comment]
@sirushtim  I think we are talking about two different things here. Adding an option could be done easily with click(the library that I mentioned earlier) and parsing yaml is done through pyyaml or some other config parsing library. These two are handled separately in the scenario that I am proposing. 

http://click.pocoo.org/5/
@chemistry-sourabh I think I recommended voting here as @sirushtim and I had different ideas. 
@sirushtim , Discuss this either internally and please prioritize this as its blocking.

**Comment by sirushtim** [comment]
@ravisantoshgudimetla I think the idea that the oslo.config library offers a complementary feature to sometimes retrieve config options from the command line is being misinterpreted as an entire command line parsing library such as click. However, if you really want a different library written which does for the most bit, the same thing oslo.config does except in YAML, I'm fine with that.

Do we really need to block the progress of the other patches because of this patch? Can they not be independent?

**Comment by chemistry-sourabh** [comment]
Can someone add what we are voting for ? My guess is oslo or Yaml right ?

**Comment by chemistry-sourabh** [comment]
@apoorvemohan I recommend stepping in here as you wanted the config to be simple

**Comment by chemistry-sourabh** [comment]
I got another possible solution. I am going to change every option back to a variable and driver sections will remain as dictionaries. I think this will remove the check option issue @sirushtim was having and will have simplicity @apoorvemohan along with dicts needed for dynamic driver loading.

Please Don't say no :P

**Comment by chemistry-sourabh** [comment]
I sought of agree with @apoorvemohan who wanted a simple config. I feel that oslo is overkill, but I feel that yaml will have the simplicity and declarative syntax (Still dont understand this shit!!) which both of you guys want. If we settle on Yaml, I will update this PR with that and push.

@naved001 and @apoorvemohan please settle this tommorow as it is an holiday (due to snow)

**Comment by chemistry-sourabh** [comment]
This PR has the config related to drivers so if it is not merged before then the other Pos won't work.

**Comment by sirushtim** [comment]
@chemistry-sourabh I don't have a specific opinion on what library we should use. Whatever config library you're writing, just make sure it takes hardly a line or two to add a new config option, make it required, keep it a default, etc. That way, config will be easy to manage.

Yes, it is important that we act fast on this PR. We either ignore it and add config options related to drivers anyway or we just make the necessary changes that there's consensus on quickly and move forward.

**Comment by sirushtim** [comment]
I've explained in detail about the library API for config parsing in https://github.com/CCI-MOC/ims/pull/41#issuecomment-278816147

Let me know if you'll have any comments on it.

**Comment by chemistry-sourabh** [comment]
Closing this as PR is merged. If need to continue discussion on oslo vs yaml. Raise another issue
